Add option to Metrics/BlockNesting to actually check blocks

georgyangelov opened the pull request:
Currently, the `Metrics/BlockNesting` cop does not consider blocks to be a nesting level. There are some cases where this is desirable.

This PR makes the block nesting cop configurable. The option is called `ConsideredBlocks` and can have a value of `none`, `multiline` and `all`.

`none` is the default which does not change the cop's current behavior. `multiline` considers multiline blocks as a nesting level and `all` counts both single-line and multiline blocks.

While implementing this, I saw that the two mixins `ConfigurableMax` and `ConfigurableEnforcedStyle` cannot be used both in a cop. They were both defining a `parameter_name` method and one was overriding the other. I have changed the name of these methods to `style_parameter_name` and `max_parameter_name`.

Please verify that my approach in the block detection is correct since this is my first time editing a cop and I'm not entirely familiar with Node's interface.

P.S. I changed my GitHub username and edited that in the contributor list.

bbatsov commented:
I took a cursory glance at the code and I wonder why are single-line and multi-line blocks treated separately. After all blocks are blocks, despite of their length...

bbatsov commented on lib/rubocop/cop/metrics/block_nesting.rb:
```diff
@@ -43,9 +49,22 @@ def check_nesting_level(node, max, current_level)
           "Avoid more than #{max} levels of block nesting."
         end
+
+        def style_parameter_name
```
Seems odd to be using the `style`-related code for something that's not the cop's enforced style, but an additional config option. @jonas054 Thoughts?

georgyangelov commented:
@bbatsov, I agree. However, there may be multiple points of view regarding what is considered a nesting level. One might consider only the indented code as a nesting level (which does not include inline blocks). This is not my view on the point, but I think this will allow for more customizability and may ease the transition. If you think the distinction is not needed, I will leave it out.

bbatsov commented:
Let's not overdo this. I'd suggest treating all blocks in the same manner for starters and see what feedback this change is going to get when merged.

bbatsov commented:
ping @georgyangelov :-)

georgyangelov commented:
@bbatsov, sorry for the delay. I've updated the option and rebased the branch on the latest master.

I've left the `parameter_name` change in here because it seemed good to have. Let me know if you want to remove it from the PR.

bbatsov commented:
Thanks! 👍
